flow/environments/xsede.py

```diff
@@ -215,8 +215,69 @@ def add_args(cls, parser):
         )
 
 
+class ExpanseEnvironment(DefaultSlurmEnvironment):
+    """Environment profile for the Expanse supercomputer.
+
+    https://www.sdsc.edu/support/user_guides/expanse.html
+    """
+
+    hostname_pattern = r".*\.expanse\.sdsc\.edu$"
+    template = "expanse.sh"
+    cores_per_node = 128
+    gpus_per_node = 4
+
+    @classmethod
+    def _get_mpi_prefix(cls, operation):
+        if operation.directives["nranks"] == 0:
+            return ""
+        # We must specify the pmi2 option as the SLURM configuration of Expanse uses the
+        # special pmi2 preprocessing. See `man srun` on an Expanse login node.
+        base_prefix = "srun --mpi=pmi2"
+        rank_option = f"-n {operation.directives['nranks']} "
+        # We have to handle the case with and without OpenMP differently due to the
+        # --cpus-per-task option.
+        if operation.directives["omp_num_threads"] == 0:
+            return base_prefix + rank_option
+        cpus_per_task = f"--cpus-per-task={operation.directives['omp_num_threads']} "
+        return base_prefix + cpus_per_task + rank_option
+
+    @classmethod
+    def _get_omp_prefix(cls, operation):
+        # The current user states to run OpenMP jobs with mpirun even when not using MPI
+        base_prefix = super()._get_omp_prefix(operation)
+        if operation.directives["nranks"] == 0:
+            return base_prefix + f"mpirun -np {operation.directives['num_omp_threads']}"
```

Comment:
This seems very strange to me. `mpirun` will launch one process for each rank requested in `-np`. OpenMP programs expect just a single process with the `OMP_NUM_THREADS` environment variable set. Like in the "Basic OpenMP Job" example from the Expanse documentation.

```
#SET the number of openmp threads
export OMP_NUM_THREADS=24

#Run the job
./hello_openmp
```

Reply:
I remember thinking this was strange too, but the User documentation doesn't show this any more, so maybe I just hallucinated this, but I removed this.